src/bank.rs

```diff
@@ -246,149 +252,238 @@ impl Bank {
+    fn apply_credits(
+        tx: &Transaction,
+        accounts: &mut [Account],
+        instruction: &Instruction,
+    ) -> Result<()> {
+        match instruction {
             Instruction::NewContract(contract) => {
                 let plan = contract.plan.clone();
                 if let Some(payment) = plan.final_payment() {
-                    self.apply_payment(&payment, accounts);
+                    Self::apply_payment(&payment, &mut accounts[1]);
+                    Ok(())
+                } else if !accounts[1].userdata.is_empty() {
+                    Err(BankError::ContractAlreadyPending(tx.keys[1]))
                 } else {
-                    let mut pending = self
-                        .pending
-                        .write()
-                        .expect("'pending' write lock in apply_credits");
+                    let mut pending = HashMap::new();
                     pending.insert(tx.signature, plan);
+                    //TODO this is a temporary on demand allocaiton
+                    //until system contract requires explicit allocation of memory
+                    accounts[1].userdata = serialize(&pending).unwrap();
+                    accounts[1].tokens += contract.tokens;
+                    Ok(())
                 }
             }
             Instruction::ApplyTimestamp(dt) => {
-                let _ = self.apply_timestamp(tx.from, *dt);
+                Self::apply_timestamp(tx.keys[0], *dt, &mut accounts[1]);
+                Ok(())
             }
             Instruction::ApplySignature(signature) => {
-                let _ = self.apply_signature(tx.from, *signature);
+                Self::apply_signature(tx.keys[0], *signature, accounts);
+                Ok(())
             }
             Instruction::NewVote(_vote) => {
-                trace!("GOT VOTE! last_id={:?}", &tx.last_id.as_ref()[..8]);
                 // TODO: record the vote in the stake table...
+                trace!("GOT VOTE! last_id={}", tx.last_id);
+                Ok(())
             }
         }
     }
-    fn save_data(&self, tx: &Transaction, accounts: &mut HashMap<Pubkey, Account>) {
-        //TODO This is a temporary implementation until the full rules on memory management for
-        //smart contracts are implemented. See github issue #953
-        if !tx.userdata.is_empty() {
-            if let Some(ref mut account) = accounts.get_mut(&tx.from) {
-                if account.userdata.len() != tx.userdata.len() {
-                    account.userdata.resize(tx.userdata.len(), 0);
-                }
-                account.userdata.copy_from_slice(&tx.userdata);
+    /// Budget DSL contract interface
+    /// * tx - the transaction
+    /// * accounts[0] - The source of the tokens
+    /// * accounts[1] - The contract context.  Once the contract has been completed, the tokens can
+    /// be spent from this account .
+    pub fn process_transaction_of_budget_instruction(
```

Comment:
so potentially this function and all it's subfunctions could be moved out of bank.rs into a separate file as it really represents a single kind of contract right?  budget_contract.rs or something?